### PR TITLE
c_keyframe.c decompiled and documented

### PR DESCRIPTION
**Pull request by Thar0**
~~There's currently bss reordering issues but wanted to put this up for reviews to start.~~

### CODE REVIEW

File: include/z64keyframe.h [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+// These flags are mutually exclusive, if XLU is set OPA is not set and vice-versa.
+#define KEYFRAME_DRAW_OPA (0 << 0)
+#define KEYFRAME_DRAW_XLU (1 << 0)
+
+typedef enum {
+    KEYFRAME_NOT_DONE,
+    KEYFRAME_DONE_ONCE,
+    KEYFRAME_DONE_LOOP
+} KeyFrameDoneType;
```
Comment (AngheloAlf) [inline include/z64keyframe.h:33]:
```suggestion
typedef enum {
    /* 0 */ KEYFRAME_NOT_DONE,
    /* 1 */ KEYFRAME_DONE_ONCE,
    /* 2 */ KEYFRAME_DONE_LOOP
} KeyFrameDoneType;
```

File: src/code/c_keyframe.c [hunk 2]
```diff
@@ -1,85 +1,1330 @@
...
+ *                           The limb data contains the index to select which callback to run.
+ *
+ * @note Original name unknown
+ */
+void Keyframe_InitFlex(KFSkelAnimeFlex* kfSkelAnime, KeyFrameFlexSkeleton* skeleton, KeyFrameAnimation* animation,
+                       Vec3s* jointTable, Vec3s* morphTable, KeyframeTransformCallback* transformCallbacks) {
+    Keyframe_ResetFlex(kfSkelAnime);
+    FrameCtrl_Init(&kfSkelAnime->frameCtrl);
+    kfSkelAnime->skeleton = (KeyFrameFlexSkeleton*)Lib_SegmentedToVirtual(skeleton);
+    kfSkelAnime->animation = (KeyFrameAnimation*)Lib_SegmentedToVirtual(animation);
```
Comment (AngheloAlf) [inline src/code/c_keyframe.c:177]: This returns `void *` so there's no need for the cast here.
```suggestion
    kfSkelAnime->skeleton = Lib_SegmentedToVirtual(skeleton);
    kfSkelAnime->animation = Lib_SegmentedToVirtual(animation);
```

Could you check the other redundant casts?

File: src/code/c_keyframe.c [hunk 3]
```diff
@@ -1,85 +1,1329 @@
...
+    pos.z = jointData->z;
+
+    Matrix_Push();
+
+    newDList = limbDList = limb->dList;
+    drawFlags = limb->drawFlags;
+
+    if (overrideKeyframeDraw == NULL ||
+        (overrideKeyframeDraw != NULL &&
+         overrideKeyframeDraw(play, kfSkelAnime, *limbIndex, &newDList, &drawFlags, arg, &scale, &rot, &pos) != 0)) {
```
Comment (hensldm) [inline src/code/c_keyframe.c:675]:
```suggestion
         overrideKeyframeDraw(play, kfSkelAnime, *limbIndex, &newDList, &drawFlags, arg, &scale, &rot, &pos))) {
```
To match skelanime.c overrideLimbDraw which also just returns a boolean

File: src/code/c_keyframe.c [hunk 4]
```diff
@@ -1,85 +1,1329 @@
...
+        if ((kfSkelAnime->transformCallbacks == NULL) || (limb->callbackIndex == KF_CALLBACK_INDEX_NONE) ||
+            (kfSkelAnime->transformCallbacks[limb->callbackIndex] == NULL) ||
+            kfSkelAnime->transformCallbacks[limb->callbackIndex](play, kfSkelAnime, *limbIndex, &newDList, &drawFlags,
+                                                                 arg) != 0) {
```
Comment (hensldm) [inline src/code/c_keyframe.c:679]:
```suggestion
            kfSkelAnime->transformCallbacks[limb->callbackIndex](play, kfSkelAnime, *limbIndex, &newDList, &drawFlags,
                                                                 arg)) {
```
These should just return booleans?

File: src/code/c_keyframe.c [hunk 5]
```diff
@@ -1,85 +1,1329 @@
...
+    OPEN_DISPS(play->state.gfxCtx);
+
+    Matrix_Push();
+
+    newDList = limbDList = limb->dList;
+    drawFlags = limb->drawFlags;
+
+    if (overrideKeyframeDraw == NULL ||
+        (overrideKeyframeDraw != NULL &&
+         overrideKeyframeDraw(play, kfSkelAnime, *limbIndex, &newDList, &drawFlags, arg, &rot, &pos) != 0)) {
```
Comment (hensldm) [inline src/code/c_keyframe.c:1196]:
```suggestion
         overrideKeyframeDraw(play, kfSkelAnime, *limbIndex, &newDList, &drawFlags, arg, &rot, &pos))) {
```

**Review by hensldm — APPROVED** [review]
Just a few small nits. Also could use a format :)

**Review by AngheloAlf — APPROVED** [review]

**Pull request merged**